### DISCUSSION TITLE
Hedera Ledger Service (HLS) Suggestion

### DISCUSSION THREAD

**Discussion by mattsmithies** [hiero-ledger/hiero-improvement-proposals#131] [Ideas]
👋 Welcome all,

Firstly this is a continuation of discussion started by @jamesgao319 seen on [this discussion page](https://github.com/hashgraph/hedera-improvement-proposal/discussions/119) with high quality comments from @lucashenning and others. I have had an introductory conversation with @bmgentile on this matter.

# Enter HLS, a new native service

Enough is enough. Liquidity is fragmented, smart contracts are fragile, and the DLT ecosystem needs to be as accessibile to developers as possible.

I want this to change.

### I see a future where:

- All available liquidity is core of the Hedera network, network-level deep liquidity. 
- No more fragmented liquidity, no longer trapped into a single set of contracts deployed by a single organsation.
- Simple SDK methods for the offer and purchase of tokens.
- Every single DEX client shares the same liquidity.

### As a developer I can:

- Build any kind of non-custodial token and NFT marketplace by default.
- Hook into the native liquidity, for any token including NFTs.
- Map the offers to my users to the related "token" definitions using offchain verifications.
- Display options for purchase, routing, and set fees.
- Enable my users to offer their tokens to the world.

### As a user I have the ability to:

- Mint any token
- Offer any token to the world, in exchange for a list of token expectations, that have been associated.
- Buy any token, using the association flow.

The goal is to produce the simplest and smart contract-less experience for any developer to build true liquidity and DeFi services with Hedera.

# Expected outcome

1. I want a direct yes or no from @donaldthibeau or @lbaird whether this is something that Hedera wants, or if it would only be through a third party validator network (which I would claim as a weak position).
2. I would like help to spec out this service in the community and the hedera team, explore other features, and edge cases. Although my time is limited I would welcome video calls, ongoing editing, and writing.
3. After we have reached consensus hand to the Hedera team for building out the feature set.

I've been internally battling with this problem (Liquidity and DEXs) for around 6 months, I believe that a new approach to deep and native liquidity could be something very special to the industry.

[My previous comment](https://github.com/hashgraph/hedera-improvement-proposal/discussions/119#discussioncomment-1049777), taken verbatim.

---

As someone that effectively created an early stage DEX and Pooling mechanisms on Hedera I would suggest that leaning into the strengths and speed of the network. I would personally adjust the perspective or belief that smart contracts are even required for us, as they have limited bandwidth within the network.

The abuse and financial risk of smart contracts are high [Vitalik even commented on this matter](https://cointelegraph.com/news/ethereum-must-innovate-beyond-just-dapps-for-defi-degens-vitalik-buterin). Hedera as a network provides us with the tools to create any solution, any appnet to drive these applications. 

There are parties, like Limechain and others building bridges, this is far more immediately important than DeFi or LP pools

Due to this realization (in February) this is the primary reason why more technical projects like **Unibar** were slowed down. 

_As a solution_ I would suggest exploring a new core Hedera service along the lines of a **Ledger** or **Exchange** service. 
Let's call it **HLS** Hedera Ledger Service, it could be in theory bundled into HTS.

The purpose of **HLS** would be to effectively extend accounts and HTS to provide state to issue token offers and execute atomic swaps, at the native layer. 

> A state store for the request and swap of HTS tokens.

The flow would be as follows:

1. A user wants to "offer" a token for sale.
2. They provide a **token_id** of their token and a criteria of expected outcomes, so I would like to sell (hedera) DOV for hedera bridged Bitcoin or Ethereum or Litecoin or anything. (This could solve the omni pool problem, provide Uniswap V3-like interfaces, and provide exponentially more opportunities for price routing)
3. A buyer could swap for a token based on the criteria, that has been offered on the native layer, price doesn't matter, only the quantity requested.

## A HLS client app

On the Application layer, it would be responsible for aggregating the liquidity options and matching up the different routing scenarios.

## Build the entire thing yourself

To be honest, any team could build this entire process out as a separate process as it isn't that difficult, all it would require is a validator network, think scheduled transactions for the upcoming auction demo.

All of these designs are always a little more complicated but my fear is that "duplicating" the experience of Ethereum on Hedera isn't a good idea, Hedera is a different beast entirely with a huge amount of potential that has to be unlocked. 

Personally, I'd love to see HLS become a thing as the network would have a decentralized mechanism for providing deep liquidity, imagine something like [woo network](https://woo.network/) at the native layer driven through developer friendly SDKs or APIs.

**Comment by djlongy** [comment]
I fully support this proposal and think it is important to the future ecosystem on the Hedera network to allow for ever increasing future enterprise adoption.

**Comment by kiraraneda** [comment]
A really great idea and concept Matt. HLS will be to DeFi what HTS is to NFT's. A must have solution for Hedera, and one that must be explored.

**Comment by jamesgao319** [comment]
Great suggestion, Matt! 🙏👍

**Comment by BrianHBARbrain** [comment]
Very cool idea to have the all Dexs to share liquidity, so only arbitrage will happen CEX to DEX.  So orders are created offchain but settled on-chain? 

Smart contract AMM use LP tokens. Would this remove that altogether? Would multisig need to be involved?  Feels likeFork and Etherdelta smart contract again going away from Uniswap Sushiswap so no Liquidity providers needed just a large OTC order book system with safe on chain atomic swaps.

I’ve love to hear how the OTC AS IS with atomic swaps could tap into this deeper liquidity!

**Reply by mattsmithies** [reply]
This is the general idea for the first pass, effectively an order book to extend the atomic swap capibilities. With this said it would be lovely to have some mechanism for AMM pools and would be happy to discuss further but my worry is that we'd want a system more akin to V3's approach of price ranges for capital efficiency, to unlock single-sided liquidity.

I don't believe there is an automated/anoymous OTC system on Hedera yet, outside of the Unibar prototype.

**Reply by kiraraneda** [reply]
Hi Matt. How would you summarize this HIP and put it forward as a question? I will submit it for the next community town hall, as I'm sure others will also.

**Reply by mattsmithies** [reply]
I'd add this as a summary. 

> How does Hedera feel about expanding the atomic swap functionality to empower users and developers to create an on-ledger order book at the network level? In other words, enable anonymous value transfer to be agnostic of a given appnet.

Irrespective of whether this approach is bought forward we have a written record of the demand of a missing part of Hedera, which is very exciting.

**Comment by crypto-lorian** [comment]
I think DEXs is the most important thing we can bring thing to hedera right now to push price action.

**Comment by SonicSealion** [comment]
I'm not a developer myself, however, as a hedera supporter I believe this would be one of the biggest drivers to further ecosystem adoption for Hedera and Hbar and think it should be a top priority.

I have biz dev and design experience and would love to help any team that decides to take this on.

**Comment by Relixx84** [comment]
I am 90+% all in on Hedera and have full faith in the future of the company. The trust and security that I see/feel in Hedera is way higher than any other platforms. I would love to see this feature implemented. Innovation is one thing that really gives Hedera an advantage. I think HCS will be a vital tech in everyone's future. This will be another great feature as well!

**Comment by justin-atwell** [comment]
Overall I really like the idea of having something like HLS for smaller startups. I think Like @mattsmithies said it has a lot of potential strengths. I do have some concerns though so allow me to play the "fifth dentist":

1. Centralization (Could be a strength or weakness of course)
2. One of the largest benefits to having my own LPs is that I can immediately access it and have control over it. By having a centralized LP I introduce a bottleneck by having to go fetch it from HLS first
3. Increased Cyclomatic Complexity. If I used HLS, I would potentially need to code around a host of exceptions that I wouldn't need to had I simply held my own liquidity.
4. Race conditions. Hedera is indeed incredibly fast, but this adds a certain amount of chattiness to the overall architecture which is worth keeping in mind.

**Reply by mattsmithies** [reply]
Mr @justin-atwell I like this rebuttal a lot, I was looking forward to this.

1. I would invite you to look into [Woo Network](https://woo.org/network), if you aren't aware of the project yet, it aims to provide deep liquidity across the board to DeFi and CeFi applications, where we are different is that it allows anyone with any token to offer and buy in a simple in a democratic fashion with little to no gatekeepers. I would ask if you feel if Hedera is a centralized solution for consensus and tokens, what is the line in the sand for you?
2. I would love to see LPs on Hedera but I'm not sure if it would be possible from a reg point of view, let alone reaching consensus on the AMM algorithm the network would run on. Let's take Uniswap for instance the AMM approach is significantly different between V2 and V3. In turn, V3 is more capital efficient and this is effectively trying to push in that direction but with the ability to ask for multiple tokens I'd love to introduce the ability for price ranges and possibly LP/Receipt tokens. [I've attached my LP design of Unibar](https://cdn.discordapp.com/attachments/808599253840166912/808599276284280872/UNIBAR_one-sided_liquidity_pools.png), to give an example of where my head has been since January.
3. Complexity, this is true but it could also be a non-issue, it all depends on the internal implementation of the feature set. To be honest, Hedera isn't currently the best for exception handling as there are already many status codes for handling different states for any given transaction.
4. Racey race time. Oh, this is a spicy one and a strong point. I am certain that this probably could lean either way, I know that one of Hedera's KPIs is to continue to increase TPS across the network but with that said a service such as this provides a higher risk of collisions and things of that nature, particularly in a distributed network.

Thank you for the feedback.

**Reply by justin-atwell** [reply]
Yeah those are great points. I didn't know about woo network. that's pretty cool.

**Comment by donaldthibeau** [comment]
@mattsmithies thanks for the great suggestion! It touches on some requirements accelerating in the ecosystem and is a well thought out potential option.

We are diving into some of the details including discussion with the relevant Council committees as to how these requirements could be met.

Thanks again!

**Reply by mattsmithies** [reply]
I love you @donaldthibeau. Thank you. 🤩

**Reply by walter-hernandez** [reply]
Hello @donaldthibeau any updates about how the discussion with the relevant Council committees went?

**Comment by brocxrace** [comment]
I really love this idea. I'm pretty swamped, but if you need someone to work on this I'd be interested in doing what I can.

**Reply by mattsmithies** [reply]
Right now, I believe we've made a good amount of noise and have garnered the attention of key players. Thank you.

**Comment by parabhairava** [comment]
I'm thrilled that @mattsmithies has stepped up and made an official proposal and voicing the hopes and wishes of many community members. The infrastructure for DeFi as we know it today should be a core service on Hedera -- users, and in this case, stake owners of the network shouldn't feel forced to use a layer 2 solution like Limechain while the technical implementation on as a core service could potentially overshadow the fragmented infrastructure in the broader crypto ecosystem which now spans through multiple chains, layers, with many different bridges. 

Adoption of DLT for everyday use-cases at the end-user & enterprise business level is certainly exciting, but let us not forget that cryptocurrency is money and that users of the Hedera network need trust-minimized protocols and applications that allow them to manage their money. But even more so, developers and decentralized communities need a platform that will allow them to continue innovating and bringing us one inch closer to what tomorrow has to bring.

As a community member, I can only hope that the core team, as well as many enthusiasts on the periphery, recognize how important this proposal is for us.

**Comment by csquare319** [comment]
Decentralized exchanges keep popping up , but so far none have been able to overtake Uniswap as the top decentralized exchange. Which is a shame, because Uniswap is certainly not perfect. If there were an exchange that could rectify the problems that Uniswap has rather than copying them, maybe we would see some evolution in the space.

Well now that’s a distinct possibility as the DODO Exchange is looking to take market share from Uniswap by actually improving on its problems. It’s doing that be attempting to eradicate impermanent loss, and improve liquidity by using a unique market making algorithm.

That unique new market maker algorithm has been coined Proactive Market Maker (PMM) by the folks at DODO Exchange. It is meant to replace the Automated Market maker (AMM) model that’s used by Uniswap and most other decentralized exchanges. The PMM model is capable of responding to liquidity constraints in real time, as well as adapting to changing market conditions.

article link: 
https://platoaistream.net/plato-data/dodo-exchange-review-proactive-market-making-dex/

--------

Above is a synopsis on PPM, a better market making mechanism than that is used by Uniswap and its clones. Perhaps Hedera team can look into PPM and see if it can leverage these innovations to develop its liquidity service. 

Additional links one might find useful:

https://dodoex.github.io/docs/docs/pmm/
https://learn.woo.org/products/liquidity-as-a-service
https://medium.com/woonetwork/partnership-series-wootrades-layer-1-liquidity-solution-for-exchanges-278a808ab261
